neo3/api/unwrap.py

```diff
@@ -0,0 +1,164 @@
+
+    Raises:
+        ValueError: if the index is out of range, or the value cannot be converted to an int
+    """
+    return item(res, idx).as_int()
+
+
+def as_uint160(res: noderpc.ExecutionResult, idx: int = 0) -> types.UInt160:
+    """
+    Convert the stack item at `idx` to an UInt160
+
+    Args:
+        res:
+        idx: the index in the result stack to fetch the stack item from.
+
+    Raises:
+        ValueError: if the index is out of range, or the value cannot be converted to an UInt160
+    """
+    return item(res, idx).as_uint160()
+
+
+def as_uint256(res: noderpc.ExecutionResult, idx: int = 0) -> types.UInt256:
+    """
+    Convert the stack item at `idx` to an UInt256
+
+    Args:
+        res:
+        idx: the index in the result stack to fetch the stack item from.
+
+    Raises:
+        ValueError: if the index is out of range, or the value cannot be converted to an UInt256
+    """
+
+    return item(res, idx).as_uint256()
+
+
+def as_address(res: noderpc.ExecutionResult, idx: int = 0) -> str:
+    """
+    Convert the stack item at `idx` to a NEO3 address
+
+    Args:
+        res:
+        idx: the index in the result stack to fetch the stack item from.
+
+    Raises:
+        ValueError: if the index is out of range, or the value cannot be converted to a NEO3 address
+    """
+    return item(res, idx).as_address()
+
+
+def as_public_key(res: noderpc.ExecutionResult, idx: int = 0) -> cryptography.ECPoint:
+    """
+    Convert the stack item at `idx` to a public key
+
+    Args:
+        res:
+        idx: the index in the result stack to fetch the stack item from.
+
+    Raises:
+        ValueError: if the index is out of range, or the value cannot be converted to a bool
+        ECCException: if the resulting key is not valid on the SECP256R1 curve
+    """
+    return item(res, idx).as_public_key()
+
+
+def as_list(res: noderpc.ExecutionResult, idx: int = 0) -> list:
+    """
+    Convert the stack item at `idx` to a list
+
+    Args:
+        res:
+        idx: the index in the result stack to fetch the stack item from.
+
+    Raises:
+        ValueError: if the index is out of range, or the value cannot be converted to a list
+    """
+    return item(res, idx).as_list()
+
+
+def as_dict(res: noderpc.ExecutionResult, idx: int = 0) -> dict:
+    """
+    Convert the stack item at `idx` to a dictionary
+
+    Args:
+        res:
+        idx: idx: the index in the result stack to fetch the stack item from.
+
+    Returns:
+
+    """
+    return item(res, idx).as_dict()
+
+
+def item(res: noderpc.ExecutionResult, idx: int = 0) -> noderpc.StackItem:
+    """
+    Fetch the stack item at `idx` from the result stack. Performs basic validation and bounds checking
+
+    Args:
+        res:
+        idx: the index in the result stack to fetch the stack item from.
+    """
+    check_state_ok(res)
+    if idx > len(res.stack) - 1:
+        raise ValueError("Too few result items")
+    return res.stack[idx]
```

Comment:
I saw that this method `item `returns a `StackItem`, but the other methods use the `ExecutionResult`. How is the conversion done? It seems possible to run a transaction and get multiple results from it, but it's not clear to me how

Reply:
What other methods are we talking about? All functions in `unwrap` take `ExecutionResult`  as input. Internally they call a similarly named function on the `StackItem` that is returned.

Pseudo; this
```python
result: ExecutionResult = Object()  # assume this is a real result
stack_item = unwrap.item(result, 0)  # checks VM state is HALT and get the stack item at index 0 from the `ResultStack`
my_int = stack_item.as_int()

```
is the same as
```python
result: ExecutionResult = Object()
my_int = unwrap.as_int(result)
```

The implementation of the unwrap functions used to do a bit more. But I moved that to the `StackItem` class itself so you can also convert from there (in case you prefer to work with the `ResultStack` yourself)